dimod/variables.py

```diff
@@ -99,16 +99,27 @@ def count(self, v):
         return int(v in self)
 
     def relabel(self, mapping):
-        try:
+
+        if PY2:
             old_labels = set(mapping)
-            new_labels = set(mapping.values())
-        except TypeError:
-            raise ValueError("mapping targets must be hashable objects")
+
+            # this try-catch is duplicated below so that when python 2 is
+            # deprecated this one can be removed
+            try:
+                new_labels = set(mapping.itervalues())
+            except TypeError:
+                # when new labels are not hashable
+                raise ValueError("mapping targets must be hashable objects")
```

Comment:
In 2.7 `mapping.values()` returns a list, we do lots of `v in new_labels` which would be O(n) for a list.

Reply:
I have changed `set(mapping)` to `mapping`

Reply:
I'm surprised by this change of `set(mapping)` -> `mapping`.

What I think is happening:
I thought mapping was a dict (in both Python 2 and 3). By calling `old_labels = set(mapping)`, I thought we were grabbing the `mapping` _keys_, putting them in a set, and then assigning them to `old_labels`. 

Confusion:
Why should `old_labels` get assigned to the entire `mapping` dict?

Reply:
The requirement is that `old_labels` be an iterable over the old labels with fast `v in old_labels` checks. In python3 we use `mapping.keys()` to make explicit that we only care about the keys and because in python3, `.keys()` returns a _view_ rather than a new object (which would mean a copy). The dict itself also qualifies, but is less explicit because it has some other stuff (the keys/new labels) hanging around. But since in python2, `.keys()` returns a list, which fails the " fast `v in old_labels` " criteria, we go with the less explicit option.

Reply:
I have moved things around and added a comment to hopefully make it clearer.

Reply:
Thanks! You've made things clear.

Reply:
Note, I was wrong about `mapping.values()` having fast lookup in python3, it is actually `O(n)` because you need to iterate over the dict. I have fixed it in the latest (and hopefully last) commit.